embed history/privacy policy/license in app

obfusk opened the pull request:
Fixes #1324.

- [x] added raw resources:
  - [x] HTML copy of GPL
  - [x] copy CHANGELOG.md
  - [x] copy PRIVACY.md
    - [ ] CatimaLoyalty/Website#26
- [x] show dialogs with HTML
- [x] basic markdown parsing for changelog & privacy policy
- [x] fix `readTextFile` to read the whole file (seems fine for `contributors.txt`)

---

<details>
<summary>screenshots</summary>

![Screenshot_20230606_154607](https://github.com/CatimaLoyalty/Android/assets/1260687/7f0c10d9-bf54-48f8-91c3-57696f9329f7)
![Screenshot_20230606_154644](https://github.com/CatimaLoyalty/Android/assets/1260687/55b7fef8-724a-428b-9555-5837370668ec)
![Screenshot_20230606_155319](https://github.com/CatimaLoyalty/Android/assets/1260687/08796c2b-c6f2-48f1-883b-7921dd1e0d6d)

</details>

TheLastProject commented:
Oh and the files Gradle creates should be added to the .gitignore

obfusk commented:
> Oh and the files Gradle creates should be added to the .gitignore

I added a gitignore in the folder they are copied to.

TheLastProject commented:
> I added a gitignore in the folder they are copied to.

I don't know if that's a good idea, Gradle may try to put that file into the .apk

obfusk commented:
> It feels silly that "Version History" has "Changelog" as first word. It feels like a duplicate title

Agreed. Was waiting for feedback before cleaning up minor stuff like that. But I can remove that.

> Why is there an extra license file? Was the LICENSE file in the repo root not good enough?

Because the HTML version looks nicer :)

> Privacy Policy also has Privacy Policy as one of the first words, feels duplicate too

Can fix that too.

> The list entries look kinda broken? like, no space between the dot and text somehow. Wonder why that's happening >.<

Yeah that bothered me too. But I don't know how to fix the HTML rendering :/

I can add an `&nbsp;` to the beginning of the list item, but that'll make the second line misaligned when it wraps.

> The scrolling through the dialog feels awful, like super heavy, but I doubt that's fixable, probably some stupid Android issue

I didn't notice that. Emulator is always laggy. Will try on device.

> We should use the fastlane instead of the CHANGELOG because it's translated but let's leave that for another MR

Good point. But yeah, seems like a good idea to improve that in a separate PR.

obfusk commented:
> I don't know if that's a good idea, Gradle may try to put that file into the .apk

It's not in the debug APK. But I can move it if you prefer.

obfusk commented:
<details>
<summary>screenshot with space</summary>

![Screenshot_20230608_211156](https://github.com/CatimaLoyalty/Android/assets/1260687/ff222ad5-cc8f-491f-a525-80190ca2b63f)

</details>

obfusk commented:
> The scrolling through the dialog feels awful, like super heavy, but I doubt that's fixable, probably some stupid Android issue

Oof. Yeah. Will see if I can fix.

obfusk commented:
Looks like I can make scrolling smooth by putting the `TextView` inside a `ScrollView` instead of making it scollable. But then links are no longer clickable. Can't find a solution to get both :/

obfusk commented:
Made the other changes (removed duplicated headers, added the space in lists).

obfusk commented:
How about just adding a link to the online version to the top of each dialog? That way users have the option to view it in the browser like before, avoiding the scrolling issues. I think that's a decent compromise given we can't have all three of kinetic scrolling, clickable links, and consistent UI (WebView is going to be hard to style).

obfusk commented:
I made some changes:

- [x] added links to the online version to all dialogs
- [x] fixed scrolling using a custom `onTouch` event handler + `ScrollView`
- [x] added & used `Utils.linkify()` because we still can't use autolinking if we want kinetic scrolling

<details>
<summary>screenshots</summary>

![Screenshot_20230610_200606](https://github.com/CatimaLoyalty/Android/assets/1260687/c25b03f3-f24d-4fce-b071-631fcbc54912)
![Screenshot_20230610_200618](https://github.com/CatimaLoyalty/Android/assets/1260687/7bed5757-847a-4cf8-aa72-8bd3b91f2a3c)
![Screenshot_20230610_200626](https://github.com/CatimaLoyalty/Android/assets/1260687/2d8e2128-b49e-4840-8100-7185331c53d5)

</details>

TheLastProject commented:
I need to review this again later but I'm thinking that the "View online" text is probably better implemented as a `setNeutralButton` with as text "View Online". That way the button becomes translatable too.

obfusk commented:
Makes sense. Though of course none of the text linked to is translated.

I can update the PR for that, unless you'd like to review the rest first?

TheLastProject commented:
Let's do the change first, it's not that much to review anyway I think :)

obfusk commented:
Ack. Rebased on `main`. Will make the changes.

obfusk commented:
LGTM. Do we need to update the "last changed" in the privacy policy?

TheLastProject commented:
> LGTM. Do we need to update the "last changed" in the privacy policy?

Why? I don't think we changed anything?

obfusk commented:
> Why? I don't think we changed anything?

I made a small change: 0829c401ec5ced7f51fcef9ededf8db553f95590

TheLastProject commented on PRIVACY.md:
```diff
@@ -0,0 +1,21 @@
+Catima offers a feature to share cards with other users. All the relevant data is in the generated shareable URLs and never transmitted to our servers. When viewed through catima.app, the data in the URL is rendered using client-side Javascript to further ensure no data is ever transmitted to us.
+
+# Changes
+This Privacy Policy may be updated from time to time for any reason. We will notify you of any changes to our Privacy Policy by posting the new Privacy Policy on the Catima website. You can view the Privacy Policy as of the time that version of the app was published directly in the app itself, though this may not reflect later changes to the online version. You are advised to consult this Privacy ...
```
```suggestion
This Privacy Policy may be updated from time to time for any reason. We will notify you of any changes to our Privacy Policy by posting the new Privacy Policy to https://catima.app/privacy-policy/. A snapshot of the Privacy Policy is available within the Catima app, however, it may be outdated. When the Privacy Policy on the website and in the app differ, the website should be considered leading. You are advised to consult the Privacy Policy regularly for any changes, as continued use is deemed approval of all changes.
```

(Also don't think "You can check the history of this policy by checking this history of this file in GitHub." actually means anything to... most users)

obfusk replied:
I'm not entirely sure how to handle it. And Google may disagree either way.

To me, it seems odd to have the online version be leading as the app is completely offline and any changes are likely to reflect changes to the app itself, unless a mistake was corrected.

The version shipped with the app should reflect what the code of the app itself does. And that should be the only thing that's relevant. There is no "backend" that could have changed.

Otherwise agreed.

obfusk replied:
I also really dislike stuff like "You are advised to consult the Privacy Policy regularly for any changes, as continued use is deemed approval of all changes", but not sure how to handle it better. I guess we could inform users of updates now that it's embedded in the app. But that's out of scope either way.

TheLastProject replied:
Honestly, the privacy policy literally comes down to "I collect as little data as I possible can". It's a formality, I wouldn't even have written one if not for Google requiring it.

The main goal of the privacy policy is to satisfy Google and cover myself. And it just seems safer to be able to update stuff online, I don't want to end up in a situation where I have to rush out an app update (and risk breaking things) just for a privacy policy fix of any kind.

If this was a real company, yeah, this would irk me too. But right now it's doing the most of preventing drama with a very limited resource: my spare time.

obfusk replied:
Ack. I wanted to point that out, but I certainly see no need to spend a lot of time on this. I will update with your suggested changes.

TheLastProject commented on PRIVACY.md:
```diff
@@ -0,0 +1,21 @@
+**Last updated**  
+January 27 2023
```
```suggestion
August 30 2023
```

Guess we should update it yeah

obfusk replied:
Will fix.

obfusk commented on PRIVACY.md:
```diff
@@ -0,0 +1,21 @@
+
+# Contact us
+If you have any questions regarding privacy while using the Application, or have questions about our practices, please contact us via email at catima.g9ex3@hackerchick.me.
+
+# Latest version
+The latest version of this privacy policy can be found online at https://catima.app/privacy-policy/
```
It's more explicit this way, but fair enough. Will fix.
